pkg/util/active_groups.go

```diff
@@ -91,11 +84,31 @@ func (ag *ActiveGroups) PurgeInactiveGroupsForUser(userID string, deadline int64
 	return deletedGroups
 }
 
+func (ag *ActiveGroups) PurgeInactiveGroups(inactiveTimeout time.Duration, cleanupFuncs ...func(string, string)) {
+	userIDs := make([]string, len(ag.timestampsPerUser))
```

Comment:
You can't access `ag.timestampsPerUser` outside the lock. Please move it after `ag.mu.RLock()`.

Reply:
Also, to simplify the code, initialise the capacity, not the length, and then just use `userIDs = append(userIDs, userID)`. This removes the need of keeping `i`.